topics/icrc_28_trusted_origins.md

```diff
@@ -0,0 +1,161 @@
+# ICRC-28: Trusted Origins
+
+[![Status Badge](https://img.shields.io/badge/STATUS-DRAFT-ffcc00.svg)](https://github.com/orgs/dfinity/projects/31)
+[![Standard Issue](https://img.shields.io/badge/ISSUE-ICRC--28-blue?logo=github)](https://github.com/dfinity/wg-identity-authentication/issues/115)
+
+<!-- TOC -->
+* [ICRC-28: Trusted Origins](#icrc-28-trusted-origins)
+  * [Summary](#summary)
+  * [Origin](#origin)
+  * [Get trusted origins](#get-trusted-origins)
+    * [Canister metadata](#canister-metadata)
+    * [Canister method](#canister-method)
+  * [Use-Cases](#use-cases)
+    * [Hot Signer Use-Case](#hot-signer-use-case)
+    * [Cold Signer Use-Case](#cold-signer-use-case)
+<!-- TOC -->
+
+## Summary
+
+Relying parties wants to request delegations from signers that can make authenticated calls to their own canisters,
+while at the same time still requiring user approval to make authenticated calls to canisters it doesn't own.
+
+**Example**
+
+- DSCVR requests a delegation that can make authenticated calls to its own canisters: dscvr_can_1 and dscvr_can_2
+- When DSCVR wants the user to transfer tokens, it will need to request the user's approval in a signer prompt
+
+For a signer to verify if a relying party should receive a delegation for a given list of canister targets,
+it will need to check for each canister target if the relying party is trusted.
+
+## Origin
+
+The window origin (as defined in https://developer.mozilla.org/en-US/docs/Glossary/Origin) is used to identify one
+relying party from another.
+
+> The origin is the concatenation of the protocol and "://", the host name if one exists, and ":" followed by a port
+> number if a port is present and differs from the default port for the given protocol. Examples of typical origins
+> are https://example.org (implying port 443), http://example.net (implying port 80), and http://example.com:8080.
+
+The origin can't be trusted with some transport layers. It's up to the signer to implement a trustable way to identify
+the origin of the relying party.
+
+The origin **MUST** be identified in a trustable way by the signer for example through `postMessage`.
+
+## Get trusted origins
+
+A canister **MUST** make both a metadata entry and a canister method available, both **MUST** return the same list of
+trusted origins.
+
+### Canister metadata
```

Comment:
Exposing this via canister metadata might not be ideal for reusability of wasm modules. I.e. embedding the information in canister metadata makes it a property of the module rather than its configuration.

For this reason, I would advocate to make this a query call and mandating certification.

In the initial proposal @dostro has pushed back against using query calls with certification due to the complexity imposed on the implementer. However, [libraries](https://github.com/dfinity/response-verification/) have improved greatly since then so IMHO this point is being addressed already by making certification easier.

Reply:
Certification is still a bit tricky to implement correctly when you start mixing http response certification and custom things like a list of trusted origins query response. So ideally I would like to see a certified query call, but then it's certification and verification needs to be standardized (standardized certified data approach for query calls).

But I agree that using metadata would make it non dynamic data which could block use cases where you would want to make it dynamic (can't think of any right now).

Reply:
Updated to use an update call. I would avoid caching, a signer could run the calls for each canister target in parallel so I don't think there's a need for caching. 

Also caching would make the spec more complicated than needed. I would expect a new delegation chain to be requested before the current delegation chain is expired, that in combination with it happening in the background without user interaction means it taking a few seconds does not cause any impact on the user experience.